core/src/main/scala/org/http4s/Query.scala

```diff
@@ -23,29 +23,37 @@ import scala.collection.immutable
   * When rendered, the resulting `String` will have the pairs separated
   * by '&' while the key is separated from the value with '='
   */
-final class Query private (val pairs: Vector[KeyValue]) extends QueryOps with Renderable {
+final class Query private (value: Either[Vector[KeyValue], String])
+    extends QueryOps
+    with Renderable {
+  lazy val pairs: Vector[KeyValue] = value.fold(identity, Query.parse)
```

Comment:
I proposed a lazy val last night.  We could also consider the null init pattern used in loggers, which is a bit faster when there is no contention.